### PR TITLE
Masked fused CE loss

### PR DESCRIPTION
**Pull request by Datta0**
Fixes : https://github.com/unslothai/unsloth/issues/5230

Respect masking when calculating chunked loss. Only accumulate if the chunk has non masked tokens

### CODE REVIEW

**Comment by danielhanchen** [comment]
Confirmed this hits the loop in unsloth_zoo/fused_losses/cross_entropy_loss.py where a chunk with every label at -100 goes through the mean reduction path and returns NaN into the accumulated loss and grads. The direction is right, but the per chunk .item() adds a host sync on every forward even when nothing is masked. Could you gather the per chunk counts in a single sync instead?

File: unsloth_zoo/fused_losses/tests.py [hunk 1]
```diff
@@ -19,30 +19,31 @@
...
+    devices = ["cuda", "cuda", "cuda", "cuda", "cuda",],
+    scalings = [1.0, 16.0, 2.0, 4.0, 1.0,],
+    logit_scale_multiplys = [0.0, 1.0, 0.0, 0.0, 0.0,],
+    logit_scale_divides = [0.0, 0.0, 2.0, 0.0, 0.0,],
+    logit_softcappings = [10.0, 0.0, 0.0, 50.0, 0.0,],
+    seeds = [3407, 3408, 3409, 3410, 5230,],
+    mask_ratios = [0.0, 0.1, 0.2, 0.3, 0.0,],
+    suffix_masks = [False, False, False, False, True,],
+    lm_head_requires_grads = [False, False, True, True, False,],
+    lm_bias_requires_grads = [False, True, False, True, False,],
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/fused_losses/tests.py]: Correct, and it was a real miss: at c1b3aaac lm_head_requires_grads and lm_bias_requires_grads had 4 entries against 5 for bszs and suffix_masks, so zip() truncated and the new suffix case never ran.

Fixed at head. All the lists are now 6 long, and the extra case is a suffix mask with a TRAINABLE head and bias, so gradient accumulation across skipped chunks is covered too, which the original 5th case (both frozen) did not reach.

File: unsloth_zoo/fused_losses/cross_entropy_loss.py [hunk 2]
```diff
@@ -325,6 +325,19 @@ def accumulate_chunk(
         global _FUSED_CE_COMPILE_SUPPORTED
         uncompiled_accumulate_chunk = accumulate_chunk
 
+        chunks = []
+        for grad_inputs_j, hidden_states_j, labels_j in zip(__grad_inputs, __shift_states, __shift_labels):
+            if bool((labels_j != -100).any().item()):
+                chunks.append((grad_inputs_j, hidden_states_j, labels_j,))
+            else:
+                grad_inputs_j.zero_()
+        pass
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py:393]: Correct, and it was worse than a stall: measured on one B200 (bsz 4 x 2048 seq, hd 1024, vocab 32768, bf16, three interleaved rounds), a .item() per chunk made this a 2.4x to 2.7x REGRESSION on dense batches and it did not reliably pay for itself even on sparse ones.

    case      chunks  empty   main   per-chunk .item()   head
    dense        128      0   41.1               98.3    42.1
    dense        512      0   77.3              210.6    81.3
    block11      512    452   76.5               78.5    14.8
    suffix32     512    500   74.5               47.0     8.1

Fixed at head, but not with the batched stack in the suggestion: torch.stack([... .any() for ...]) is still one reduction kernel per chunk, and get_chunk_size routinely picks hundreds. Folding the validity mask to one flag per chunk is a fixed number of kernels and one sync regardless of chunk count, giving 3 to 5 percent on dense and 5x to 9x on sparse.

File: unsloth_zoo/fused_losses/cross_entropy_loss.py [hunk 3]
```diff
@@ -341,7 +354,7 @@ def accumulate_chunk(
         if _FUSED_CE_COMPILE_SUPPORTED is None and torch_compile and \
             accumulate_chunk is not uncompiled_accumulate_chunk:
```
[the remark this answers is not in the document]
Comment (danielhanchen) [reply unsloth_zoo/fused_losses/cross_entropy_loss.py:415]: Correct at the time. c1b3aaac really did have 'except StopIteration: _FUSED_CE_COMPILE_SUPPORTED = True', which latched the flag without ever running the compiled path, so a later failure would have skipped the eager fallback.

Resolved a different way at head: the all-empty case now returns before the probe block is reached, so next(_iter) cannot raise and the flag is left as None for the next non-empty batch to probe. Verified: an all-masked batch followed by a real one leaves compile_state None after the empty call and True after the real one, with the real batch matching an unchunked reference.

**Comment by danielhanchen** [comment]
@codex review

**Comment by danielhanchen** [comment]
@codex security review

**Comment by danielhanchen** [comment]
Checked this against the loop in unsloth_zoo/fused_losses/cross_entropy_loss.py, and the three earlier points are all addressed at head: the mask is folded to one flag per chunk with a single sync, the suffix case now really runs with a trainable head and bias, and the all-empty batch returns before the compile probe so the flag stays unlatched. Will get this reviewed.

**Comment by danielhanchen** [comment]
Closing this one. Thanks for the work on it. The reason is what the testing turned up rather than anything wrong with the implementation itself, so recording it here in full.

**1. It does not fix #5230.** Losses are bit-identical to main across 480 configs and 2880 numeric checks, and an end-to-end run with a sparse last-turn mask learns the same thing either way (LoRA deltas 2497.0 / 2498.9 / 2497.7). The sparse-mask issue is real, but this is not its cause.

**2. The divide-by-zero it does fix is mostly already guarded.** `_filter_fully_masked` in `dataset_utils.py` drops fully masked rows before training, and it runs after tokenisation and truncation, so standard text SFT cannot produce an all-ignored microbatch. What is left unguarded is the vision and OCR collators, which rebuild labels at collate time, and streaming datasets, which the filter explicitly cannot scan. That is a narrow surface.

**3. The chunk skip only pays above 8 chunks.** `torch.chunk` splits contiguously, so at 4 chunks or fewer every chunk still contains its own row's trainable tail and nothing is skippable. Fraction of empty chunks for a 2048-token sequence:

| chunks | 11% answer tail | 1.6% answer tail |
|---|---|---|
| 4 | 0/4 (0%) | 0/4 (0%) |
| 8 | 4/8 (50%) | 4/8 (50%) |
| 32 | 28/32 (88%) | 28/32 (88%) |
| 128 | 112/128 (88%) | 124/128 (97%) |

Chunk count comes from free GPU memory, so the benefit is zero on a roomy card and large on a tight one. Real, but not broad enough to justify the extra branch in the hot path right now.

One thing worth keeping if this is ever revisited: as originally submitted the per-chunk `.item()` made the **dense** path 2.4x to 2.7x slower, since each call forces a host sync. Measured at a 7B-ish shape, dense went from 38.8ms to 109.2ms. Folding the mask check into a single kernel brings it back to 41.8ms while keeping the sparse win, and that version is on the branch.
